tt_metal/api/device.hpp

```diff
@@ -0,0 +1,305 @@
+
+/**
+ * @brief Retrieves the number of DRAM channels on the device.
+ *
+ * @param device The device to query.
+ * @return Number of DRAM channels.
+ */
+int GetNumDramChannels(Device device);
+
+/**
+ * @brief Retrieves the available L1 size per worker core on the device.
+ *
+ * @param device The device to query.
+ * @return L1 size per core in bytes.
+ */
+uint32_t GetL1SizePerCore(Device device);
+
+/**
+ * @brief Computes the storage grid size for the device.
+ *
+ * @param device The device to query.
+ * @return CoreCoord representing the storage grid size.
+ */
+CoreCoord GetComputeWithStorageGridSize(Device device);
+
+/**
+ * @brief Retrieves the DRAM grid size for the device.
+ *
+ * @param device The device to query.
+ * @return CoreCoord representing the DRAM grid size.
+ */
+CoreCoord GetDramGridSize(Device device);
+
+/**
+ * @brief Converts a logical core coordinate to a physical core coordinate.
+ *
+ * @param device The device to query.
+ * @param logical_core The logical core coordinate.
+ * @param core_type The type of the core.
+ * @return CoreCoord representing the physical core coordinate.
+ */
+CoreCoord PhysicalCoreFromLogical(Device device, const CoreCoord &logical_core, const CoreType &core_type);
+
+/**
+ * @brief Retrieves the worker core coordinate corresponding to a logical core.
+ *
+ * @param device The device to query.
+ * @param logical_core The logical core coordinate.
+ * @return CoreCoord representing the worker core coordinate.
+ */
+CoreCoord WorkerCoreFromLogical(Device device, const CoreCoord &logical_core);
+
+/**
+ * @brief Retrieves the Ethernet core coordinate corresponding to a logical core.
+ *
+ * @param device The device to query.
+ * @param logical_core The logical core coordinate.
+ * @return CoreCoord representing the Ethernet core coordinate.
+ */
+CoreCoord EthernetCoreFromLogical(Device device, const CoreCoord &logical_core);
+
+/**
+ * @brief Enables the program cache on the device.
+ *
+ * @param device The device to modify.
+ */
+void EnableProgramCache(Device device);
+
+/**
+ * @brief Disables and clears the program cache on the device.
+ *
+ * @param device The device to modify.
+ */
+void DisableAndClearProgramCache(Device device);
+
+/**
+ * @brief Pushes a work function onto the device's work queue.
+ *
+ * @param device The device to which the work will be pushed.
+ * @param work The work function to execute.
+ * @param blocking Indicates whether the operation should be blocking (default: false).
+ */
+void PushWork(Device device, std::function<void()> &&work, bool blocking = false);
+
+/**
+ * @brief Pushes a shared work function onto the device's work queue.
+ *
+ * @param device The device to which the work will be pushed.
+ * @param work Shared pointer to the work function to execute.
+ * @param blocking Indicates whether the operation should be blocking (default: false).
+ */
+void PushWork(Device device, std::function<void()> work, bool blocking = false);
```

Comment:
Consider `std::move_only_function`. I've been bitten many times by `std::function`. It **requires the function to be copy-able**. One major hurdle is you can't put a `unique_ptr` in a lambda's capture list. Nor any references.

Reply:
Thanks for this suggestion, and the example with `unique_ptr`, that would be an important restriction to recognize. However, I don't believe reference captures don't delete copy-construction, can you elaborate what you meant by that? https://godbolt.org/z/dn7a9GzrY

Reply:
Also `std::move_only_function` is C++23, and I don't believe upgrading to that standard is currently under consideration.